keps/sig-node/4112-passdown-resources-to-cri/README.md

```diff
@@ -0,0 +1,1297 @@
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The CRI runtime lacks visibility to the application resource requirements.
+
+First, the resources required by the containers of a pod are not visible at the
+pod sandbox creation time. This can be problematic for example in the case of
+VM-based runtimes where all resources need to be reserved/prepared when the VM
+(i.e. sandbox) is being created.
+
+Second, the kubelet does not provide complete information about the container
+resources specification of native and extended resources (requests and limits)
+to CRI. However, various use cases have been identified where detailed
+knowledge of all the resources can be utilized in container runtimes for more
+optimal resource allocation to improve application performance and reduce
+cross-application interference.
+
+This KEP proposes CRI API extensions for providing complete view of pods
+resources at sandbox creation, and, providing unobfuscated information about
+the resource requests and limits to container runtimes.
+
+## Motivation
+
+When the pod sandbox is created, the kubelet does not provide the CRI runtime
+any information about the resources (such as native resources, host devices,
+mounts, CDI devices etc) that will be required by the application. The CRI
+runtime only becomes aware of the resources piece by piece when containers of
+the pod are created (one-by-one).
```

Comment:
Which of this information must come from the Kubelet, and which can be read from the k8s API? Resource requests & limits and volumes can all be read directly from the pod object. Do they need to be plumbed through the CRI, or can you just read from the k8s API instead?

Reply:
Adding to that, for mounts we want more/different information than what is available on the Kubernetes API level. At least the `host_path` that will be mounted to the container.

Reply:
Additional to the other points, in Kata we cannot handle `subPath` in `volumeMounts` because subPath is handled completely by the kubelet. https://github.com/kubernetes/kubernetes/blob/d7774fce9a7fcec890d7c0beffacd6ae34152b01/pkg/volume/util/subpath/subpath_linux.go#L175